### ISSUE TITLE
Class constructor FileManager cannot be invoked without 'new'.

### ISSUE THREAD

**Issue by micooz** [less/less.js#3414]
It seems that **v3.10.0** breaks my build, webpack logs below:

```
ERROR in ./src/pages/score/components/current/no-join/index.less
Module build failed (from ./node_modules/_mini-css-extract-plugin@0.5.0@mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/_less-loader@4.1.0@less-loader/dist/cjs.js):


Class constructor FileManager cannot be invoked without 'new'
      in undefined (line undefined, column undefined)
    at runLoaders (/home/admin/build/node_modules/_webpack@4.39.2@webpack/lib/NormalModule.js:313:20)
    at /home/admin/build/node_modules/_loader-runner@2.4.0@loader-runner/lib/LoaderRunner.js:367:11
    at /home/admin/build/node_modules/_loader-runner@2.4.0@loader-runner/lib/LoaderRunner.js:233:18
    at context.callback (/home/admin/build/node_modules/_loader-runner@2.4.0@loader-runner/lib/LoaderRunner.js:111:13)
```

Everything is ok when downgrade to **v3.9.0**, please investigate this issue and hope to fix it as soon as possible.

**Comment by xiaoningzihai** [comment]
I have the same problem，you mast change you 'package.json' => "less": "~3.9.0", limit version

**Comment by alexyan** [comment]
+1

**Comment by AlwaysSkylll** [comment]
+1

**Comment by lzf0402** [comment]
+1!
急 ～

**Comment by Liqihan** [comment]
+1

**Comment by yolo2013** [comment]
same problem

**Comment by Doneitha** [comment]
+1  version  less 3.9.0    less-loader 4.1.0

**Comment by Doneitha** [comment]
可以把package.json中 less版本的 ^ 去掉  重新安装一下 v3.9.0 应该就可以解决问题 大家可以试一下   
虽然 锁定了版本 但是由于 ^, 所以会下载3 一下的最新版本， 最新版本应该是有问题的。

**Comment by lizux** [comment]
+1

**Comment by wangmingquan** [comment]
+1

**Comment by AlwaysSkylll** [comment]
`const shortname = new this.less.FileManager().extractUrlParts(filename).filename;` 

这个类似代码引起的？

**Comment by Tindtily** [comment]
omg

**Comment by starzle** [comment]
Same here. npm picks up this "minor" release automatically, and breaks all of my builds. :(

**Comment by huxiaoqi567** [comment]
+1

**Comment by smackgg** [comment]
v3.10.0 same error

**Comment by AlixWang** [comment]
+1

**Comment by doxiaodong** [comment]
hurry

**Comment by simple7** [comment]
+1

**Comment by lixiaohulee** [comment]
just use less-loader@5.0.0   every thing will be ok

**Comment by 0x80** [comment]
@lixiaohulee 	unless you're not using less-loader and everything will not be ok

**Comment by wangningbo93** [comment]
+1

**Comment by asiareal** [comment]
+1

**Comment by KatDenisenko** [comment]
I have the same problem (((

**Comment by Lalitha-Iyer** [comment]
+1

**Comment by kikeztw** [comment]
i have the same problem :/

**Comment by matthew-dean** [comment]
This is an issue with an older version of `less-loader`, not Less, which did not inherit the Less filemanager correctly. See: https://github.com/less/less.js/issues/3414#issuecomment-522223041

**Comment by lulinglm** [comment]
thanks

**Comment by ZhiGang-Fang** [comment]
解决了吗?

**Comment by 670755937** [comment]
> 解决了吗?

你可以，把你package.js 里的less依赖，重新安装最近小版本依赖。
![image](https://user-images.githubusercontent.com/27992771/63221969-af9be300-c1d3-11e9-80ac-804847abc93e.png)

**Comment by khom-outsourcify** [comment]
+1

**Comment by aleen42** [comment]
Oooooooooooops!!!

**Comment by montanaflynn** [comment]
We're using antd which broke: https://github.com/ant-design/ant-design/issues/18343

How can we fix the issue?

**Comment by aleen42** [comment]
@montanaflynn keep a older envrionment and generate locked dependencies map with `npm-shrinkmap.json` may be a good choice.

**Comment by montanaflynn** [comment]
@aleen42 thanks, I hope this can be fixed soon, I'm not very familiar with all the npm stuff.

**Comment by aleen42** [comment]
@montanaflynn I have met many times when we broke during building due to the unlocked dependencies, including sub dependencies. This is a confusing problems for most FE developers.

**Comment by montanaflynn** [comment]
@aleen42 we fixed it by adding 

```
  "devDependencies": {
    "less-loader": "^5.0.0"
  }
```

**Comment by 1312543912** [comment]
Why does this problem arise? 
Node.js version is not right?

**Comment by TengSimon** [comment]
It's not a node version issue.

**Comment by TengSimon** [comment]
1

**Comment by TengSimon** [comment]
> Why does this problem arise?
> Node.js version is not right?

+1

**Comment by aleen42** [comment]
@montanaflynn you can also fix the problem by specifying a sub version for `less-loader`:

```json
devDepedencies: {
    "less-loader": "4.1.0",
    "less": "3.9.0"
}
```

**Comment by TecHunter** [comment]
Upgrade with 
`yarn upgrade less@^3.10.1 less-loader@^5.0.0`

**Comment by gozeon** [comment]
use `"less": "~3.9.0"`

**Comment by LizhangX** [comment]
+1

**Comment by jenniferneale** [comment]
+1

**Comment by matthew-dean** [comment]
Incidentally, for those watching this thread, I'll re-open and probably submit a fix later today for older `less-loader` versions, just to have things be less annoying.

It's one of those edge cases where actually using the `Class` syntax to replace a prototype function throws an error without `new`. A FileManager is supposed to always be invoked with `new` (that's the documented API), and older versions of `less-loader` apparently did not.

**Comment by aleen42** [comment]
I think that it is better for `less-loader` to downgrade its dependencies below `3.9.0`.

**Comment by matthew-dean** [comment]
@aleen42 Why would that be necessary, when the latest version of `less-loader` works fine?

**Comment by aleen42** [comment]
Assume that some projects may not need to take care the risk of high level upgrading from `4.x` to `5.x`.

**Comment by LeonWuV** [comment]
+666

**Comment by matthew-dean** [comment]
Can someone try: https://github.com/less/less.js/pull/3421?

Branch is at: https://github.com/matthew-dean/less.js/tree/release/v3.10.2

The challenge in testing this is that `less-loader`'s own tests fail on 3.9.0, so it's not clear what's expected. But, with Less's exported classes transpiled down to functions, the same errors at least get triggered in `less-loader`, so it's not clear if that's expected output or not. It's also not clear _how_ or why this error occurs in Less loader v4, and doesn't occur in v5, except it may be a Webpack issue on v4?

In any case, if people can confirm that this branch as a Less dependency doesn't cause the class issue, then I'll merge it in.

**Comment by wantwant123** [comment]
+1 version less 3.9.0 less-loader 4.1.0

**Comment by MustafinVadim** [comment]
+1

**Comment by resyin** [comment]
+1

**Comment by lang1427** [comment]
Manually specify less version   ：
npm i less@3.9.0 -D     
ok!!!

**Comment by JunlinZhu-Tommy** [comment]
Fix by downgrade to 3.9.0

**Comment by vipcxj** [comment]
Just before, I delete the node_modules and type `npm i`, this error happen. So the latest version `3.10.1` still has this bug. I choose downgrade to 3.9.0

**Comment by ezgitek** [comment]
How would I fix this bug if I am using @zeit/next-less? Any help appreciated.

**Comment by hua1995116** [comment]
@ezgitek  
```
npm i @zeit/next-less
npm i less@3.9.0
```

**Comment by buerxixi** [comment]
https://forum.vuejs.org/t/vue-cli/72008/7
"less": "^3.10.1",
"less-loader": "^5.0.0",
is running

**Comment by matthew-dean** [comment]
> No one with less-loader v4 issues seemed willing to test [3.10.2] 🤷‍♂, but in my testing, Less-loader's v4 tests broke differently with 3.10.1 (class constructor issue), and broke the same with this PR. So..... not sure if it fixes it, but it should be no worse?

I merged that possible fix into master (but not yet released to NPM). Still would love for someone to confirm instead of the +1's.

**Comment by jenniferneale** [comment]
I can confirm I just tried the current master branch of less locally with less-loader ^4.1.0, and my project is able to build again. 👍

**Comment by jenniferneale** [comment]
Thanks for your work on this!

**Comment by matthew-dean** [comment]
@jenniferneale Thanks for confirming! I'll publish to NPM later today then, can't do it from work.

**Comment by pod4g** [comment]
+1

**Comment by H4ngman** [comment]
Is this issue fixed with 3.10.3?

**Comment by matthew-dean** [comment]
@H4ngman Should be. Re-open if not.
